opensearch-operator/pkg/reconcilers/upgrade.go

```diff
@@ -0,0 +1,363 @@
+		Client: client,
+		ResourceReconciler: reconciler.NewReconcilerWith(client,
+			append(opts, reconciler.WithLog(log.FromContext(ctx).WithValues("reconciler", "scaler")))...),
+		ctx:               ctx,
+		recorder:          recorder,
+		reconcilerContext: reconcilerContext,
+		instance:          instance,
+	}
+}
+
+func (r *UpgradeReconciler) Reconcile() (ctrl.Result, error) {
+	// If versions are in sync do nothing
+	if r.instance.Spec.General.Version == r.instance.Status.Version {
+		return ctrl.Result{}, nil
+	}
+
+	lg := log.FromContext(r.ctx)
+
+	// If version validation fails log a warning and do nothing
+	if err := r.validateUpgrade(); err != nil {
+		lg.V(1).Error(err, "version validation failed", "currentVersion", r.instance.Status.Version, "requestedVersion", r.instance.Spec.General.Version)
+		return ctrl.Result{}, err
+	}
+
+	// If there is work to do create an Opensearch Client
+	username, password, err := helpers.UsernameAndPassword(r.ctx, r.Client, r.instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	clusterClient, err := services.NewOsClusterClient(fmt.Sprintf("https://%s.%s:9200", r.instance.Spec.General.ServiceName, r.instance.Namespace), username, password)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	r.osClient = clusterClient
+
+	//Fetch the working nodepool
+	nodePool, currentStatus := r.findWorkingNodePool()
+
+	// Work on the current nodepool as appropriate
+	switch currentStatus.Status {
+	case "Pending":
+		// Set it to upgrading and requeue
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			currentStatus.Status = "Upgrading"
+			r.instance.Status.ComponentsStatus = append(r.instance.Status.ComponentsStatus, currentStatus)
+			return r.Status().Update(r.ctx, r.instance)
+		})
+		r.recorder.Eventf(r.instance, "Normal", "upgrading", "beginning upgrade of data node pool %s", currentStatus.Component)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 15 * time.Second,
+		}, err
+	case "NonDataPending":
+		// Set it to upgrading and requeue
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			currentStatus.Status = "UntrackedUpgrade"
+			r.instance.Status.ComponentsStatus = append(r.instance.Status.ComponentsStatus, currentStatus)
+			return r.Status().Update(r.ctx, r.instance)
+		})
+		r.recorder.Eventf(r.instance, "Normal", "upgrading", "beginning rollout of non data node pool %s", currentStatus.Component)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 15 * time.Second,
+		}, err
+	case "Upgrading":
+		err := r.doDataNodeUpgrade(nodePool)
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 30 * time.Second,
+		}, err
+	case "Finished":
+		// Cleanup status after successful upgrade
```

Comment:
add test to that behaviour

Reply:
I've added some additional tests by manipulating the status.